### PR TITLE
Fix missing environment when yielding an array

### PR DESCRIPTION
**Pull request by sklam**
Fixes issue #1254

### CODE REVIEW

**Comment by seibert** [comment]
Looks like this is failing on Python 2 on Linux and Mac.

**Comment by sklam** [comment]
Looks like the problem is due to the libc symbol `puts` is undefined in pycc module.
Either remove the `puts` call that help inform about the fatal error about missing Environment or update PYCC to link with the appropriate library.

**Comment by pitrou** [comment]
What is the error?

**Comment by seibert** [comment]
This seems to be working (just Python 3.4 is pending on Travis-CI).  If that passes, is this ready to merge?

File: numba/pythonapi.py [hunk 1]
```diff
@@ -143,8 +144,18 @@ def __init__(self, context, builder):
         self.gil_state = Type.int(_helperlib.py_gil_state_size * 8)
         self.py_buffer_t = ir.ArrayType(ir.IntType(8), _helperlib.py_buffer_size)
 
-    def get_env_manager(self, env, env_body):
-        return EnvironmentManager(self, env, env_body)
+    def get_env_manager(self, env, env_body, env_ptr):
+        is_null = cgutils.is_null(self.builder, env_ptr)
```
Comment (pitrou) [inline numba/pythonapi.py]: I don't think this method should emit any LLVM code at all. Can't you put the test in one of the callers (if at all necessary)?

**Comment by sklam** [comment]
Here's a list of tests that fails if we don't have the fix and an exception is raised when an EnvironmentManager is created but the environment pointer is NULL:

```
======================================================================
ERROR: test_dump_ir_generator (numba.tests.test_debug.TestGeneratorDebugOutput)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_debug.py", line 133, in test_dump_ir_generator
    out = self.compile_simple_gen()
  File "/Users/sklam/dev/numba/numba/tests/test_debug.py", line 128, in compile_simple_gen
    self.assertPreciseEqual(list(cres.entry_point(2, 5)), [2, 5])
RuntimeError: missing Environment

======================================================================
ERROR: test_issue_1254 (numba.tests.test_generators.TestGeneratorWithNRT)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 404, in test_issue_1254
    outputs = list(random_directions(5))
RuntimeError: missing Environment

======================================================================
ERROR: test_gen1 (numba.tests.test_generators.TestGenerators)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 123, in test_gen1
    self.check_gen1()
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 120, in check_gen1
    self.check_generator(pygen, cgen)
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 106, in check_generator
    self.assertEqual(next(cgen), next(pygen))
RuntimeError: missing Environment

======================================================================
ERROR: test_gen2 (numba.tests.test_generators.TestGenerators)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 136, in test_gen2
    self.check_gen2()
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 133, in check_gen2
    self.check_generator(pygen, cgen)
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 106, in check_generator
    self.assertEqual(next(cgen), next(pygen))
RuntimeError: missing Environment

======================================================================
ERROR: test_gen3 (numba.tests.test_generators.TestGenerators)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 149, in test_gen3
    self.check_gen3()
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 146, in check_gen3
    self.check_generator(pygen, cgen)
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 106, in check_generator
    self.assertEqual(next(cgen), next(pygen))
RuntimeError: missing Environment

======================================================================
ERROR: test_gen4 (numba.tests.test_generators.TestGenerators)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 162, in test_gen4
    self.check_gen4()
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 159, in check_gen4
    self.check_generator(pygen, cgen)
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 106, in check_generator
    self.assertEqual(next(cgen), next(pygen))
RuntimeError: missing Environment

======================================================================
ERROR: test_gen6 (numba.tests.test_generators.TestGenerators)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 190, in test_gen6
    self.check_gen6()
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 186, in check_gen6
    l.append(next(cgen))
RuntimeError: missing Environment

======================================================================
ERROR: test_gen7 (numba.tests.test_generators.TestGenerators)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 205, in test_gen7
    self.check_gen7()
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 202, in check_gen7
    self.check_generator(pygen, cgen)
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 106, in check_generator
    self.assertEqual(next(cgen), next(pygen))
RuntimeError: missing Environment

======================================================================
ERROR: test_ndenumerate (numba.tests.test_generators.TestGenerators)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 261, in test_ndenumerate
    self.check_np_flat(gen_ndenumerate)
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 248, in check_np_flat
    self.check_generator(pyfunc(arr), cr.entry_point(arr))
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 106, in check_generator
    self.assertEqual(next(cgen), next(pygen))
RuntimeError: missing Environment

======================================================================
ERROR: test_ndindex (numba.tests.test_generators.TestGenerators)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 239, in test_ndindex
    self.check_ndindex()
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 236, in check_ndindex
    self.check_generator(pygen, cgen)
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 106, in check_generator
    self.assertEqual(next(cgen), next(pygen))
RuntimeError: missing Environment

======================================================================
ERROR: test_np_flat (numba.tests.test_generators.TestGenerators)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 255, in test_np_flat
    self.check_np_flat(gen_flat)
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 248, in check_np_flat
    self.check_generator(pyfunc(arr), cr.entry_point(arr))
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 106, in check_generator
    self.assertEqual(next(cgen), next(pygen))
RuntimeError: missing Environment

======================================================================
ERROR: test_combine_gen0_gen1 (numba.tests.test_generators.TestNrtArrayGen)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 333, in test_combine_gen0_gen1
    self.test_nrt_gen0()
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 296, in test_nrt_gen0
    c_res = list(cgen(c_ary))
RuntimeError: missing Environment

======================================================================
ERROR: test_nrt_gen0 (numba.tests.test_generators.TestNrtArrayGen)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 296, in test_nrt_gen0
    c_res = list(cgen(c_ary))
RuntimeError: missing Environment

======================================================================
ERROR: test_nrt_gen0_stop_iteration (numba.tests.test_generators.TestNrtArrayGen)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 350, in test_nrt_gen0_stop_iteration
    c_res = next(c_iter)
RuntimeError: missing Environment

======================================================================
ERROR: test_nrt_gen1 (numba.tests.test_generators.TestNrtArrayGen)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_generators.py", line 315, in test_nrt_gen1
    c_res = list(cgen(c_ary1, c_ary2))
RuntimeError: missing Environment

----------------------------------------------------------------------
Ran 6661 tests in 460.127s

FAILED (errors=15, skipped=31, expected failures=3)

```

**Comment by pitrou** [comment]
This looks ok to me. Holding on Travis-CI...

**Comment by pitrou** [comment]
There are timeout errors on the Python 2 builds.

**Comment by sklam** [comment]
Interestingly, PYCC on python2 has a NULL value for _dynfunc.Closure.

**Comment by sklam** [comment]
BTW, disassembling with lldb save the day

**Pull request merged**